### PR TITLE
DATAUP-263 cell stuck in sending

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes

This addresses the bug where an app cell can appear to get stuck right after running with the message "Sending...". The only option is either to cancel the cell (which might not work), or reload the Narrative (which might result in lost work).

The core of the error involves the FSM behind the App Cell getting conflicting or out of order messages. To solve it, this change updates the FSM to allow progression from the "execution-requested" state to any of the running states or post-run error states, incase those messages get mangled. There's a chance that some intermediate message could show up afterward, but this would likely just be ignored in favor of the updated job state.

Other changes include cleanup and linting to the `appCell.js` and `appCellWidget.js` modules and some testing.

Testing this particular change is difficult as it requires some deep integration to exist. The easiest way I found to replicate the issue is to start running an app and immediately go to the `Kernel` menu and it `Restart`.

# Jira Ticket / Issue #
https://kbase-jira.atlassian.net/browse/DATAUP-263
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
* Not really automatically tested. That might warrant its own ticket, issue.
* Start an app cell (not bulk import)
* Immediate go to the Kernel menu and it restart
* It should sit in "Sending..." for a while, then proceed with the app run
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: nbextensions/appCell2/appCell.js [hunk 1]
```diff
@@ -31,28 +31,33 @@ define([
         div = t('div');
 
     function isAppCell(cell) {
-        if (cell.cell_type !== 'code') {
+        if (cell.cell_type !== 'code' || !cell.metadata.kbase) {
```
Comment (briehl) [inline nbextensions/appCell2/appCell.js:34]: Some simplification here.

File: nbextensions/appCell2/main.js [hunk 2]
```diff
@@ -17,14 +17,11 @@ define([
...
     'common/runtime',
-    'common/clock',
     'common/error',
-    'kb_service/utils',
-    'kb_service/client/workspace',
     './appCell',
     'bootstrap',
     'custom/custom',
-], ($, Jupyter, Promise, Runtime, Clock, Error, serviceUtils, Workspace, AppCell) => {
+], ($, Jupyter, Promise, Runtime, Error, AppCell) => {
```
Comment (briehl) [inline nbextensions/appCell2/main.js:24]: Removed unused dependencies.

File: nbextensions/appCell2/widgets/appCellWidget.js [hunk 3]
```diff
@@ -147,17 +147,51 @@ define([
...
-            controlBarTabs = {},
             selectedJobId,
             readOnly = false,
-            viewOnly = false;
+            viewOnly = false,
+            saveTimer = null,
+            jobListeners = [],
+            // Track whether the user has selected a tab.
+            // This is reset when the user closes a tab.
+            userSelectedTab = false;
```
Comment (briehl) [inline nbextensions/appCell2/widgets/appCellWidget.js:194]: Also moved up from below, along with a few others.

File: test/unit/spec/nbextensions/appCell2/appCell-spec.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+define([
```
Comment (briehl) [inline test/unit/spec/nbextensions/appCell2/appCell-spec.js:1]: Tests for the core app cell component.

**Comment by briehl** [comment]
Some other changes here that would be nice to have.

1. Fully test the appCellWidget. I started to work on that, but there's a pile of dangling Promises that may or may not get properly resolved in the various state changes in that widget. Really, I fell into a deep hole for a couple of days working on that, until I found that my (incomplete) changes were causing problems, so I stopped and reverted it all. Sigh.
2. Automatically test the solution to this bug. I'm not sure the best way to do that with unit tests, as they all rely on having a consistent way to test the appCellWidget module and, well, ☝️

File: test/unit/spec/nbextensions/appCell2/appCell-spec.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            });
+
+            it('makes a new AppCell from scratch', () => {
+                const fakeCell = Mocks.buildMockCell('code');
+                const appCell = AppCell.make({ cell: fakeCell });
+                const appTag = 'release';
+                const appType = 'app';
+
+                // mock NMS.get_method_spec, NMS.get_method_full_info, Catalog.get_exec_aggr_stats
+                return appCell.upgradeToAppCell(TestAppSpec, appTag, appType);
```
Comment (ialarmedalien) [inline test/unit/spec/nbextensions/appCell2/appCell-spec.js:36]: does this test anything?

Comment (briehl) [reply test/unit/spec/nbextensions/appCell2/appCell-spec.js]: @ialarmedalien fleshed out the remaining tests in that module.

File: test/unit/spec/nbextensions/appCell2/appCell-spec.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                        simple_string: null
+                    },
+                    'user-settings': {
+                        showCodeInputArea: false
+                    }
+                };
+                const appCell = AppCell.make({ cell: fakeAppCell });
+                return appCell.setupCell()
+                    .then(() => {
+                        expect(appCell).toBeDefined();
```
[the remark this answers is not in the document]
Comment (briehl) [reply test/unit/spec/nbextensions/appCell2/appCell-spec.js:124]: And, yeah, looking at it now, that's a silly, useless `expect`. I'll add something better.
This is what happens when you're trying to juggle 3 tasks at once... 🤹‍♂️

[the remark this answers is not in the document]
Comment (ialarmedalien) [reply test/unit/spec/nbextensions/appCell2/appCell-spec.js:124]: 😖

**Review by ialarmedalien — APPROVED** [review]
LGTM

**Pull request merged**